backends/p4tools/common/core/target.cpp

```diff
@@ -48,23 +47,95 @@ bool Target::init(std::string deviceName, std::string archName) {
     return false;
 }
 
-bool Target::setDevice(std::string deviceName) {
-    std::transform(deviceName.begin(), deviceName.end(), deviceName.begin(), ::tolower);
-    if (defaultArchByDevice.count(deviceName) == 0U) {
+std::optional<ICompileContext *> Target::initializeTarget(std::string_view toolName,
+                                                          std::string_view target,
+                                                          std::string_view arch) {
+    // Establish a dummy compilation context so that we can use ::error to report errors while
+    // processing target and arch.
+    class DummyCompileContext : public BaseCompileContext {
+    } dummyContext;
+    AutoCompileContext autoDummyContext(&dummyContext);
+    if (!P4Tools::Target::setDevice(target)) {
+        ::P4::error("Unsupported device: %s", target);
+        return std::nullopt;
+    }
+    if (!P4Tools::Target::setArch(arch)) {
+        ::P4::error("Unsupported architecture: %s", arch);
```

Comment:
```suggestion
        error("Unsupported architecture: %s", arch);

Reply:
I will replace all these instances in a separate PR.